parsl/config.py

```diff
@@ -92,9 +92,12 @@ def __init__(self,
                  monitoring: Optional[MonitoringHub] = None,
                  usage_tracking: bool = False,
                  initialize_logging: bool = True) -> None:
+        self._executors: Sequence[ParslExecutor]
         if executors is None:
-            executors = [ThreadPoolExecutor()]
-        self.executors = executors
+            self._executors = [ThreadPoolExecutor()]
+        else:
+            self._validate_executors(executors)
+            self._executors = executors
```

Comment:
Stylistically, you might consider removing the if-condition for less mental load while reading.  Perhaps:
```python
if executors is None:
    executors = [ThreadPoolExecutor()]
assert executors is not None  # possibly necessary for mypy -- didn't run this ...
self._validate_executors(executors)
self._executors = executors
```
Now there's a single point of authority for the typing (from `__init__`'s signature), a single assignment, and it's clear that `self._executors` can't be `None`.

Alternatively, is it an error if `executors` was specified as an empty sequence?  Could use a naked truthy check to cover both:
```python
if not executors:
    executors = [ThreadPoolExecutor()]
# or possibly:
# executors = executors or [ThreadPoolExecutor()]
```

Reply:
> ... it was always a bug to replace the executor list in a config once parsl has been configured, ...

you might consider also making a copy of what's passed in.  That is, nominally we don't want a user to change (accidentally) the list they passed in, right?  Perhaps:
```python
def __init__(
    self,
        # note: Sequence -> Iterable
    executors: Optional[Iterable[ParslExecutor]] = None,
    ...
):
    executors = tuple(executors)
    if not executors:
        executors = (ThreadPoolExecutor(),)
    self._validate_executors(executors)
    self._executors = executors
    ...
```

Reply:
bbc3ddae7710775cad34e6340113610b6805d8f0 broadly makes the changes you suggested.

I think there are some weird corner cases though because of the interaction between being a bool and being an iterable:

in the presence of an executors structure that is boolean false, it won't be iterated, and a default will be selected.

in the presence of an executors structure that is boolean true but iterates to empty, a default will not be selected and a configuration error for "no executors!" will be raised.

I don't think anything forbids that second case from happening (but i'm not entirely familar with the specifications for bool vs iterable). But it's also something I'm not too fussed about  - the "obvious" structures a user might use here of tuple, list and set have the right iteration vs bool properties at least.